File: swift/Ice/filesystem/Sources/Server/MFile.swift
// Copyright (c) ZeroC, Inc.

import Ice

/// Provides an in-memory implementation of the Slice interface File.
/// We use an actor since this servant has mutable state (the file contents).
actor MFile: File {
    private let node: MNode // reuse the MNode implementation
    private var lines: [String] = []

    init(name: String) {
        self.node = MNode(name: name)
    }

    // Implements Node.name by forwarding to node.
    func name(current: Ice.Current) -> String {
        node.name(current: current)
    }

    // Implements File.read.
    func read(current _: Ice.Current) -> [String] {
        lines
    }

    // Implements File.write.
    func write(text: [String], current _: Ice.Current) {
        writeDirect(text: text)
    }

    // Writes directly to this file, without going through an Ice operation.
    func writeDirect(text: [String]) {
        lines = text
    }
}

File: java/Ice/filesystem/server/src/main/java/com/example/ice/filesystem/server/MFile.java
// Copyright (c) ZeroC, Inc.

package com.example.ice.filesystem.server;

import com.zeroc.Ice.Current;

/**
 * Provides an in-memory implementation of the Slice interface File.
 */
public class MFile extends Node implements File {
    private String[] _lines = {};

    /**
     * Constructs an {@code MFile}.
     * @param name the name of this file
     */
    MFile(String name)
    {
        super(name);
    }

    @Override
    public String[] read(Current current)
    {
        return _lines;
    }

    @Override
    public void write(String[] text, Current current) throws WriteException
    {
        writeDirect(text);
    }

    /**
     * Writes directly to this file, without going through an Ice operation.
     * @param text the text to write
     */
    void writeDirect(String[] text)
    {
        _lines = text;
    }
}
